### DISCUSSION TITLE
BUP image and UEFI Capsule generation questions

### DISCUSSION THREAD

**Discussion by igoropaniuk** [OE4T/meta-tegra#1492] [Q&A]
I've been learning recently about how BUP/UEFI capsules are generated in the latest master.
Correct my if I'm wrong:

BUP creation logic is implemented in `classes/image_types_tegra.bbclass`, in `create_bup_payload_image()`:

```
775 create_bup_payload_image() {                                                    
776     local type="$1"                                                             
777     oe_make_bup_payload ${IMGDEPLOYDIR}/${IMAGE_NAME}${IMAGE_NAME_SUFFIX}.${type}
778     for f in ${WORKDIR}/bup-payload/*_only_payload; do                          
779         [ -e $f ] || continue                                                   
780         sfx=$(basename $f _payload)                                             
781         install -m 0644 $f ${IMGDEPLOYDIR}/${IMAGE_NAME}.$sfx.bup-payload       
782         ln -sf ${IMAGE_NAME}.$sfx.bup-payload ${IMGDEPLOYDIR}/${IMAGE_LINK_NAME}.$sfx.bup-payload
783     done                                                                        
784 }      
```

This function is defined in CONVERSION_CMD for `bup-payload` image type `CONVERSION_CMD:bup-payload = "create_bup_payload_image ${type}"`, which implies that we should have some image with bup-payload type defined in `INITRAMFS_FSTYPES`.

Final UEFI capsule is generated by `tegra-uefi-capsule` recipe, which expects **BUP** blobs to be deployed to `$DEPLOY_DIR_IMAGE` in advance.

Was there any particular reason why BUP creation is done via `CONVERSION_CMD` and not in a dedicated recipe (like it's  currently done for UEFI capsule generation in`tegra-uefi-capsule` recipe)? is there any some sort of update process which requires a running initrd?

**Comment by madisongh** [comment]
There's some history to the way this is structured that pre-dates UEFI and capsules.  There's also some history to the assumption that the kernel uses a initrd.  Going forward, I'd like to get rid of the separate BUP generation and move that into the capsule creation recipes.

If you're having problems building/using a kernel without an initrd because a kernel BUP never gets generated, that should probably be fixable in the `kernel-bup-payload` recipe, which is the non-CONVERSION_CMD way to generate a kernel BUP.  The logic in there now looks like only works for bundled initramfs images, not for plain kernels with no initrd/initramfs at all.
